File: Object-Oriented-Programming/Lab9Undo/Lab9Undo/consola.cpp
#include "consola.h"
#include "Film.h"
#include <iostream>
#include <string>

using std::cin;
using std::cout;

constexpr int Optiune_Adaugare_Film = 1;
constexpr int Optiune_Stergere_Film = 2;
constexpr int Optiune_Modificare_Film = 3;
constexpr int Optiune_Afisare_Filme = 4;
constexpr int Optiune_Cautare_Film = 5;
constexpr int Optiune_Filtrare_Film = 6;
constexpr int Optiune_Sortare_Film = 7;
constexpr int Optiune_Adaugare_In_Cos_Inchirieri = 8;
constexpr int Optiune_Stergere_Din_Cos_Inchirieri = 9;
constexpr int Optiune_Golire_Cos_Inchirieri = 10;
constexpr int Optiune_Umple_Random_Cos_Inchirieri = 11;
constexpr int Optiune_Afisare_Cos_Inchirieri = 12;
constexpr int Optiune_Exportare_In_Fisier_CSV = 13;
constexpr int Optiune_Undo = 14;
constexpr int Optiune_Iesire = 0;

void UI::ui_adaugaFilm() {
	int anAparitie;
	string titluFilm, genFilm, actorPrincipal;
	cin.ignore();
	cout << "Introduceti titlul filmului: ";
	getline(cin, titluFilm);
	cout << "Introduceti genul filmului: ";
	getline(cin, genFilm);
	cout << "Introduceti anul aparitiei filmului: ";
	cin >> anAparitie;
	cin.ignore();
	cout << "Introduceti numele actorului care joaca in rolul principal: ";
	getline(cin, actorPrincipal);
	service.adaugaFilm(titluFilm, genFilm, anAparitie, actorPrincipal);
	cout << "Filmul " << titluFilm << " a fost adaugat cu succes!\n";
}

void UI::ui_stergeFilm() {
	string titluFilm;
	cin.ignore();
	cout << "Introduceti titlul filmului pentru stergere: ";
	getline(cin, titluFilm);
	service.stergeFilm(titluFilm);
	cout << "Filmul " << titluFilm << " a fost sters cu succes!\n";
}

void UI::ui_modificaFilm() {
	int anAparitie;
	string titluFilm, genFilm, actorPrincipal;
	cin.ignore();
	cout << "Introduceti titlul filmului pentru modificat: ";
	getline(cin, titluFilm);
	cout << "Introduceti noul gen al filmului: ";
	getline(cin, genFilm);
	cout << "Introduceti noul an al aparitiei filmului: ";
	cin >> anAparitie;
	cin.ignore();
	cout << "Introduceti numele actorului care joaca in rolul principal: ";
	getline(cin, actorPrincipal);
	service.modificaFilm(titluFilm, genFilm, anAparitie, actorPrincipal);
	cout << "Filmul " << titluFilm << " a fost modificat cu succes!\n";
}

void UI::ui_afiseazaFilme(const vector<Film>& filme) {
	if (filme.size() == 0) {
		cout << "\nNu exista filme in lista!\n";
		return;
	}
	cout << "Filme: \n";
	cout << "=================================\n";
	int i = 0;
	for (const auto& film : filme) {
		i++;
		cout << i << ". " << film.getTitluFilm() << "; gen: " << film.getGenFilm() << "; aparut in anul: " << film.getAnAparitie() << "; Actor in rolul principal: " << film.getActorPrincipal() << "\n";
	}
	cout << "=================================\n";
}

void UI::ui_cautaFilm() {
	string titluFilm;
	cin.ignore();
	cout << "Introduceti titlul filmului cautat: ";
	getline(cin, titluFilm);
	const Film& film = service.cautaFilm(titluFilm);
	cout << film.getTitluFilm() << "; gen: " << film.getGenFilm() << "; aparut in anul: " << film.getAnAparitie() << "; Actor in rolul principal: " << film.getActorPrincipal() << "\n";
}

void UI::ui_filtrare() {
	// 1 -> gen / 2 -> anul aparitiei / 0 -> iesire (sunt optiunile utilizatorului)
	int reperFiltrare;
	cin.ignore();
	cout << "Alegeti reperul de filtrare (1 -> gen / 2 -> anul aparitiei / 0 -> iesire): ";
	cin >> reperFiltrare;
	if (reperFiltrare == 1) {
		string genFilm;
		cout << "Introduceti genul dupa care sa se filtreze: ";
		cin.ignore();
		getline(cin, genFilm);
		ui_afiseazaFilme(service.filtreazaDupaGen(genFilm));
	}
	else if (reperFiltrare == 2) {
		int anAparitie;
		cout << "Introduceti anul dupa care sa se filtreze: ";
		cin >> anAparitie;
		ui_afiseazaFilme(service.filtreazaDupaAnAparitie(anAparitie));
	}
	else if (reperFiltrare == 0) return;
	else {
		cout << "Reper de filtrare necunoscut!\n";
		ui_filtrare();
	}
}

void UI::ui_sortare() {
	// 1 -> titlu / 2 -> actor principal / 3 -> anul aparitiei + gen / 0 -> iesire (pentru reperul de sortare) -> prima optiune a utilizatorului
	// 1 -> crescator / 2 -> descrescator (pentru ordinea de sortare) -> a doua optiune a utilizatorului
	int reperSortare, ordineSortare;
	cin.ignore();
	cout << "Alegeti reperul dupa care sa se sorteze (1 -> titlu / 2 -> actor principal / 3 -> anul aparitiei + gen / 0 -> iesire): ";
	cin >> reperSortare;
	if (reperSortare == 0) return;
	if (reperSortare != 1 && reperSortare != 2 && reperSortare != 3) {
		cout << "Reper invalid! Va rugam sa introduceti 1, 2 sau 3!\n";
		ui_sortare();
	}
	cout << "Alegeti ordinea de sortare (1 -> crescator / 2 -> descrescator): ";
	cin >> ordineSortare;
	if (ordineSortare != 1 && ordineSortare != 2) {
		cout << "Ordine de sortare invalida! Va rugam sa introduceti 1 sau 2!\n";
		ui_sortare();
	}
	const vector<Film>& filme = service.sorteaza(reperSortare, ordineSortare);
	ui_afiseazaFilme(filme);
}

void UI::ui_adaugaFilmInCosInchirieri() {
	string titluFilm;
	cin.ignore();
	cout << "Introduceti titlul filmului pentru a-l adauga in cosul de inchirieri: ";
	getline(cin, titluFilm);
	ui_afiseazaFilme(service.adaugaFilmInCosInchirieri(titluFilm));
}

void UI::ui_stergeFilmDinCosInchirieri() {
	string titluFilm;
	cin.ignore();
	cout << "Introduceti titlul filmului pentru a-l sterge din cosul de inchirieri: ";
	getline(cin, titluFilm);
	ui_afiseazaFilme(service.stergeFilmDinCosInchirieri(titluFilm));
}

void UI::ui_umpleRandomCosInchirieri() {
	int dimensiune;
	cin.ignore();
	cout << "Introduceti numarul de filme de adaugat in cosul de inchirieri: ";
	cin >> dimensiune;
	ui_afiseazaFilme(service.umpleRandomCosInchirieri(dimensiune));
}

void UI::ui_exportaInFisier() {
	string numeFisier, formatFisier;
	cin.ignore();
	cout << "Introduceti tipul de fisier in care sa se introduca cosul (CSV / HTML): ";
	getline(cin, formatFisier);
	cout << "Introduceti numele unui fisier pentru a salva continutul cosului de inchirieri: ";
	getline(cin, numeFisier);
	const vector<Film>& filme = service.getAllCosInchirieri();
	if (formatFisier == "CSV")
		service.exportaInCSV(numeFisier, filme);
	else if (formatFisier == "HTML")
		service.exportaInHTML(numeFisier, filme);
	else
		cout << "Tip fisier invalid!\n";
}

void UI::ui_undo() {
	service.undo();
	ui_afiseazaFilme(service.getAll());
}

void UI::ruleaza() {
	/*
	service.adaugaFilm("John Wick: Chapter 4", "Actiune", 2023, "Keanu Reeves");
	service.adaugaFilm("The Imitation Game", "Historic Drama", 2014, "Benedict Cumberbatch");
	service.adaugaFilm("Titanic", "Epic Romance", 1997, "Leonardo DiCaprio");
	service.adaugaFilm("Avatar 2", "Science-Fiction", 2022, "Sam Worthington");
	service.adaugaFilm("The Revenant", "American Action Drama", 2016, "Leonardo DiCaprio");
	*/
	while (true) {
		cout << "\n=================================\n";
		cout << "   Meniu inchiriere filme: \n";
		cout << "1. Adauga film.\n";
		cout << "2. Sterge film.\n";
		cout << "3. Modifica film.\n";
		cout << "4. Afiseaza toate filmele disponibile pentru inchiriere.\n";
		cout << "5. Cauta un film dupa titlu.\n";
		cout << "6. Filtreaza filme dupa titlu / anul aparitiei.\n";
		cout << "7. Sorteaza filme dupa titlu / actorul principal / anul aparitiei + gen.\n";
		cout << "8. Adauga film in cosul de inchirieri.\n";
		cout << "9. Sterge film din cosul de inchirieri.\n";
		cout << "10.Goleste cosul de inchirieri.\n";
		cout << "11.Adauga un numar de filme aleatoare in cosul de inchirieri.\n";
		cout << "12.Afiseaza cosul de inchirieri.\n";
		cout << "13.Exporta cosul in fisier CSV / HTML.\n";
		cout << "14.Undo.\n";
		cout << "0. Iesire din aplicatie.\n";
		cout << "=================================\n";
		int comanda;
		cout << ">>>";
		cin >> comanda;
		try {
			switch (comanda) {
			case Optiune_Adaugare_Film:
				ui_adaugaFilm();
				break;
			case Optiune_Stergere_Film:
				ui_stergeFilm();
				break;
			case Optiune_Modificare_Film:
				ui_modificaFilm();
				break;
			case Optiune_Afisare_Filme:
				ui_afiseazaFilme(service.getAll());
				break;
			case Optiune_Cautare_Film:
				ui_cautaFilm();
				break;
			case Optiune_Filtrare_Film:
				ui_filtrare();
				break;
			case Optiune_Sortare_Film:
				ui_sortare();
				break;
			case Optiune_Adaugare_In_Cos_Inchirieri:
				ui_adaugaFilmInCosInchirieri();
				break;
			case Optiune_Stergere_Din_Cos_Inchirieri:
				ui_stergeFilmDinCosInchirieri();
				break;
			case Optiune_Golire_Cos_Inchirieri:
				ui_afiseazaFilme(service.golesteCosInchirieri());
				break;
			case Optiune_Umple_Random_Cos_Inchirieri:
				ui_umpleRandomCosInchirieri();
				break;
			case Optiune_Afisare_Cos_Inchirieri:
				ui_afiseazaFilme(service.getAllCosInchirieri());
				break;
			case Optiune_Exportare_In_Fisier_CSV:
				ui_exportaInFisier();
				break;
			case Optiune_Undo:
				ui_undo();
				break;
			case Optiune_Iesire:
				return;
			default:
				cout << "Comanda necunoscuta!\n";
			}
			cout << '\n';
		}
		catch (const ExceptieRepo& exceptie) {
			cout << exceptie << "\n";
		}
		catch (const ExceptieValidare& exceptie) {
			cout << exceptie << "\n";
		}
	}
}

UI::~UI()
{
}

File: S1-Fundamentals-of-Programming/Simulare2/prezentare/consola.py
from erori.repository_error import RepoError
from erori.validation_error import ValidError


class UI:

    def __init__(self, service_filme, service_clienti):
        self.__service_filme = service_filme
        self.__service_clienti = service_clienti
        self.__comenzi = {
            "1": self.__meniu_filme,
            "2": self.__meniu_clienti
        }
        self.__comenzi_filme = {
            "1": self.__ui_adauga_film,
            "2": self.__ui_modifica_film,
            "3": self.__ui_sterge_film,
            "4": self.__ui_cauta_film,
            "5": self.__ui_afiseaza_filmele
        }
        self.__comenzi_clienti = {
            "1": self.__ui_adauga_client,
            "2": self.__ui_modifica_client,
            "3": self.__ui_sterge_client,
            "4": self.__ui_cauta_client,
            "5": self.__ui_afiseaza_clientii
        }

    def __ui_adauga_film(self):
        id_film = int(input("Introduceti id-ul filmului: "))
        titlu = input("Introduceti titlul filmului: ")
        descriere = input("Introduceti descrierea filmului: ")
        gen = input("Introduceti genul filmului: ")
        self.__service_filme.adauga_film(id_film, titlu, descriere, gen)
        print("Film adaugat cu succes!\n")

    def __ui_modifica_film(self):
        id_film = int(input("Introduceti id-ul filmului: "))
        titlu_nou = input("Introduceti titlul nou al filmului: ")
        descriere_noua = input("Introduceti descrierea noua a filmului: ")
        gen_nou = input("Introduceti genul nou al filmului: ")
        self.__service_filme.modifica_film(id_film, titlu_nou, descriere_noua, gen_nou)
        print("Film modificat cu succes!\n")

    def __ui_sterge_film(self):
        id_film = int(input("Introduceti id-ul filmului pentru stergere: "))
        self.__service_filme.sterge_film(id_film)
        print("Film sters cu succes!\n")

    def __ui_cauta_film(self):
        id_film = int(input("Introduceti id-ul filmului cautat: "))
        film = self.__service_filme.cauta_film(id_film)
        print(film)

    def __ui_afiseaza_filmele(self):
        filme = self.__service_filme.get_all()
        if len(filme) == 0:
            print("Nu exista filme introduse!\n")
            return
        for id_film in filme:
            print(filme[id_film])

    def __ui_adauga_client(self):
        id_client = int(input("Introduceti id-ul clientului: "))
        nume = input("Introduceti numele clientului: ")
        CNP = input("Introduceti CNP-ul clientului: ")
        self.__service_clienti.adauga_client(id_client, nume, CNP)
        print("Client adaugat cu succes!\n")

    def __ui_modifica_client(self):
        id_client = int(input("Introduceti id-ul clientului: "))
        nume_nou = input("Introduceti numele nou al clientului: ")
        CNP_nou = input("Introduceti CNP-ul nou clientului: ")
        self.__service_clienti.modifica_client(id_client, nume_nou, CNP_nou)
        print("Client modificat cu succes!\n")

    def __ui_sterge_client(self):
        id_client = int(input("Introduceti id-ul clientului: "))
        self.__service_clienti.sterge_client(id_client)
        print("Client sters cu succes!\n")

    def __ui_cauta_client(self):
        id_client = int(input("Introduceti id-ul clientului: "))
        client = self.__service_clienti.cauta_client(id_client)
        print(client)

    def __ui_afiseaza_clientii(self):
        clienti = self.__service_clienti.get_all()
        if len(clienti) == 0:
            print("Nu exista clienti introdusi!\n")
            return
        for id_client in clienti:
            print(clienti[id_client])

    def __meniu_filme(self):
        while True:
            print("   Meniu filme: ")
            print("1. Adauga film")
            print("2. Modifica film")
            print("3. Sterge film")
            print("4. Cauta film")
            print("5. Afiseaza filmele")
            print("0. Inapoi la meniul principal")
            try:
                comanda = input(">>>")
                if comanda == "":
                    continue
                if comanda == '0':
                    return
                if comanda in self.__comenzi_filme:
                    self.__comenzi_filme[comanda]()
                else:
                    print("Comanda necunoscuta!")
            except ValueError:
                print("Eroare UI: tip numeric invalid!")
            except ValidError as ve:
                print(f'Eroare Validare: {ve}')
            except RepoError as re:
                print(f"Eroare Repository: {re}")

    def __meniu_clienti(self):
        while True:
            print("   Meniul clienti: ")
            print("1. Adauga client")
            print("2. Modifica client")
            print("3. Sterge client")
            print("4. Cauta client")
            print("5. Afiseaza clientii")
            print("0. Inapoi la meniul principal")
            try:
                comanda = input(">>>")
                if comanda == "":
                    continue
                if comanda == '0':
                    return
                if comanda in self.__comenzi_clienti:
                    self.__comenzi_clienti[comanda]()
                else:
                    print("Comanda necunoscuta!")
            except ValueError:
                print("Eroare UI: tip numeric invalid!")
            except ValidError as ve:
                print(f'Eroare Validare: {ve}')
            except RepoError as re:
                print(f"Eroare Repository: {re}")

    def run(self):
        while True:
            print("   Meniul principal: ")
            print("1. Meniu gestionare filme")
            print("2. Meniu gestionare clienti")
            print("0. Iesi din program")
            try:
                comanda = input(">>>")
                if comanda == "":
                    continue
                if comanda == '0':
                    return
                if comanda in self.__comenzi:
                    self.__comenzi[comanda]()
                else:
                    print("Comanda necunoscuta!")
            except ValueError:
                print("Eroare UI: tip numeric invalid!")
            except ValidError as ve:
                print(f'Eroare Validare: {ve}')
            except RepoError as re:
                print(f"Eroare Repository: {re}")